examples/dnn_introduction3_ex.cpp

```diff
@@ -29,7 +29,7 @@ namespace model
     using net_type = loss_metric<
         fc_no_bias<128,
         avg_pool_everything<
-        typename resnet<BN>::template backbone_50<
+        typename resnet::impl<BN>::template backbone_50<
```

Comment:
What do you think about `def` or `net` instead of `impl`?  Like maybe:
```
typename resnet::def<BN>::template backbone_50
```

Reply:
Honestly, I don't have any preference. I just used `impl` because I saw that is what you used for internal implementations (although as namespace qualifiers). So, I also think either `def` or `net` would be a better name. Which one do you prefer?

Reply:
How about def then. resnet::net feels weird.

Reply:
Done!
With regards doing the tagging after each resnet level (I've found it useful for training multi-scale segmentation models), here's what I am thinking:

- adding the following tags after each level:
``` cpp
template <typename SUBNET> using rtag1 = add_tag_layer<6001, SUBNET>;
template <typename SUBNET> using rtag2 = add_tag_layer<6002, SUBNET>;
template <typename SUBNET> using rtag3 = add_tag_layer<6003, SUBNET>;
template <typename SUBNET> using rtag4 = add_tag_layer<6004, SUBNET>;
```
- adding the following skip layers, too, for convenience:
``` cpp
template <typename SUBNET> using skipr1 = add_skip_layer<rtag1, SUBNET>;
template <typename SUBNET> using skipr2 = add_skip_layer<rtag2, SUBNET>;
template <typename SUBNET> using skipr3 = add_skip_layer<rtag3, SUBNET>;
template <typename SUBNET> using skipr4 = add_skip_layer<rtag4, SUBNET>;
```
(I've chosen 6000 because ResNet has 6 letters)

Then, there comes the tricky part: once I get this model defined, I should iterate over all the layers and do:
``` cpp
layer<i>(new_net).layer_details() = layer<j>(old_net).layer_details();
```
With appropriate `i` and `j`. Once it's done, serialize the new model.

I was wondering if there was an easy way to do this (maybe with a visitor?).
But since I only need to do it once, I think it will take less time to write a macro in my editor that will generate all the assignment lines. How would you do this?
Also, I was wondering if `rtag4` should refer to the level close to the input layer or the loss layer. I've seen that you used `level4` for the one that's close to the input layer (following the same order as the number of the layers) but @reunanen used the inverse order, which makes more sense in the way I think about how the data flows through the network, but it's less consistent with the order of the layers. Sorry for asking too many questions... (:

Reply:
I'm not sure what's the best way.  Like you said, it's pretty easy to make a macro in an editor.  Like I use vim and it's super easy to do this kind of thing there.  It depends on how often you do it.  It is sort of awkward to setup something automatic that does this.  You could use a visitor that keeps an int that increments on every layer but tag layers and use that to make the mapping though.

Reply:
Ok, that's exactly what I thought, I was just wanted the expert's opinion to make sure I was not missing anything.
I guess it would be up to the end user to do it they way they want then :)